### DISCUSSION TITLE
Poll period wakeup callback function Thread SED

### DISCUSSION THREAD

**Discussion by Rajapur** [openthread/openthread#8156] [Q&A] [answered]
Hello Everyone,

I am developing thread based device (freeRTOS project) which connect to thread network.

I have configured Sleepy End Device now I have one question,

where is the callback after wakeup or poll period is over ?

I have to check is there any network Data update every wakeup/poll period is over so that if any changes then I have to re-register the SRP (border router).
if any one know how to check wake up callback let me know. Project is developed on freeRTOS.

I am using nordic SDK Thread and Zigbee and Nrf52840 DK. 

Thanks in advance 

Regards,
Rohit R

**Comment by jwhui** [comment, marked as answer]
OpenThread's SRP Client implementation should automatically track updates to the list of SRP Servers advertised in the Thread Network Data and react accordingly: https://github.com/openthread/openthread/blob/b66d521d1d08b9e8bec831a19ead00fec2a32ce9/src/core/net/srp_client.cpp#L421

If your application code really requires handling OpenThread state changes, you can register a callback using [otSetStateChangedCallback()](https://openthread.io/reference/group/api-instance#otsetstatechangedcallback).

**Reply by Rajapur** [reply]
Hi, 
My setup is I have border router running on RPI and nRF52 DK with KNX IoT stack(thread party stack). 

My job use KNX framework and send message to and from. So in KNX also doesn't have any call if there is network parameter are updated. 
So In nordic I have poll for every wakeup means if poll period is over check is there any network data are update (for example router changed or Ip address changed or port id changed) and register SRP with border router again. 

Regarding StateChangecallback it is with respect to device role changes, whether it is child role or router or leader or detached. 
I need callback when poll period is over?  in freeRTOS

Thanks and Regards
Rohit R

**Reply by jwhui** [reply]
> So In nordic I have poll for every wakeup means if poll period is over check is there any network data are update (for example router changed or Ip address changed or port id changed) and register SRP with border router again.

As previously mentioned, OpenThread's SRP Client will automatically send SRP Updates as necessary when there are state changes (e.g. IPv6 address or Thread Network Data changes): https://github.com/openthread/openthread/blob/b66d521d1d08b9e8bec831a19ead00fec2a32ce9/src/core/net/srp_client.cpp#L421

You can enable the auto host address feature using [otSrpClientEnableAutoHostAddress()](https://openthread.io/reference/group/api-srp#otsrpclientenableautohostaddress).

Can you explain why you need to send your own SRP Updates (rather than letting OpenThread handle them)?

> Regarding StateChangecallback it is with respect to device role changes, whether it is child role or router or leader or detached.
I need callback when poll period is over? in freeRTOS

The state changed callback may signal a number of different state changes. See [this enumeration](https://openthread.io/reference/group/api-instance#enumerations) for all the state changes that may be indicated.

**Reply by Rajapur** [reply]
Hi, 

> As previously mentioned, OpenThread's SRP Client will automatically send SRP Updates as necessary when there are state changes (e.g. IPv6 address or Thread Network Data changes):
> 
> https://github.com/openthread/openthread/blob/b66d521d1d08b9e8bec831a19ead00fec2a32ce9/src/core/net/srp_client.cpp#L421
> 
> You can enable the auto host address feature using [otSrpClientEnableAutoHostAddress()](https://openthread.io/reference/group/api-srp#otsrpclientenableautohostaddress).
> 
> Can you explain why you need to send your own SRP Updates (rather than letting OpenThread handle them)?

-- We are using KnX Stack and this stack build the library for coap and srp they do not contain such function or file. Which handles this notification if any changes happens. So we are using polling method to check is there in data update for every wakeup logic. 

> The state changed callback may signal a number of different state changes. See [this enumeration](https://openthread.io/reference/group/api-instance#enumerations) for all the state changes that may be indicated.
--I tried below method but i didn't get statechangecallback for network data change flag. 
the snippet, 
 static void thread_state_changed_callback(uint32_t flags, void * p_context)
{
    if (flags & OT_CHANGED_THREAD_ROLE)
    {
        switch(otThreadGetDeviceRole(p_context))
        {
            case OT_DEVICE_ROLE_CHILD:             
             break;

            case OT_DEVICE_ROLE_ROUTER:
            case OT_DEVICE_ROLE_LEADER:
                break;

            case OT_DEVICE_ROLE_DISABLED:
            case OT_DEVICE_ROLE_DETACHED:
            break;

            default:
                break;
        }

        if(flags == OT_CHANGED_THREAD_NETDATA)
        {
           srp_config_t config;

           srp_server_infor_get(chNewServerIpAddress, &ui16NewPort);
            if((memcmp(chCurrentServerIpAddress,chNewServerIpAddress,sizeof(chNewServerIpAddress))) || (ui16CurrentPort != ui16NewPort))
           {
                config.server_address        = chNewServerIpAddress;
             config.server_port           = ui16NewPort;
               memcpy(chCurrentServerIpAddress, chNewServerIpAddress, sizeof(chNewServerIpAddress));
              ui16CurrentPort = ui16NewPort;

               srp_register(15, args); //srp registration function
           }
       }
    }

    NRF_LOG_INFO("State changed! Flags: 0x%08x Current role: %d\r\n", flags, otThreadGetDeviceRole(p_context));
}
- I am new to this let me know is there anything to configure so that callback executes if any changes happens. 

Thanks and Regards
Rohit R

**Reply by jwhui** [reply]
> We are using KnX Stack and this stack build the library for coap and srp they do not contain such function or file. Which handles this notification if any changes happens. So we are using polling method to check is there in data update for every wakeup logic.

Are you using OpenThread's SRP Client implementation on an older commit? If so, I would encourage moving to a more recent OpenThread commit which includes general enhancements and bug fixes.

> I tried below method but i didn't get statechangecallback for network data change flag.

The flags are a bit mask. So you want to use something like:
```
if (flags & OT_CHANGED_THREAD_NETDATA)
```

**Reply by Rajapur** [reply]
Hi, 

Thanks for the feedback. 

> Are you using OpenThread's SRP Client implementation on an older commit? If so, I would encourage moving to a more recent OpenThread commit which includes general enhancements and bug fixes.

-- In KNX stack we are not using openthread version which has SRP support currently. And currently we are not is state to upgrade it. 

> The flags are a bit mask. So you want to use something like:
> ```
> if (flags & OT_CHANGED_THREAD_NETDATA)
> ```

--Okay. now i got it. I will make changes and test. If i disable SRP server and enable SRP server again did i get network change update in this callback and do my process. 

Thank you for the response and quick help. 

Thanks and Regards
Rohit R

**Reply by jwhui** [reply]
> In KNX stack we are not using openthread version which has SRP support currently. And currently we are not is state to upgrade it.

OK. As FYI, the OpenThread SRP Client is tightly integrated with Thread, so all of this would be handled for you already.

**Reply by Rajapur** [reply]
Hi,

> OK. As FYI, the OpenThread SRP Client is tightly integrated with Thread, so all of this would be handled for you already.
- Okay, I will discuss this with KNX team. 

And one more update from my side is, I used the openthread instant Network data changed(OT_CHANGED_THREAD_NETDATA) in statechange callback and i could able to execute my SRP logic, I am checking if any Port id or address change when we try SRP disable and SRP enable. if yes, i am registering the SRP again. 
I am testing it further and also check with team about this implementation which is acceptable or not. 

Thank you so much for help. 

Regards
Rohit R
